src/main/java/traincraft/api/SlotCraftingResult.java

```diff
@@ -0,0 +1,118 @@
+ * You can find the source code at https://github.com/Traincraft/Traincraft
+ */
+package traincraft.api;
+
+import com.google.common.collect.Lists;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.inventory.IInventory;
+import net.minecraft.inventory.InventoryCraftResult;
+import net.minecraft.inventory.InventoryCrafting;
+import net.minecraft.inventory.Slot;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.CraftingManager;
+import net.minecraft.item.crafting.IRecipe;
+import net.minecraft.util.NonNullList;
+import traincraft.blocks.trainworkbench.TrainWorkbenchRecipe;
+
+/**
+ * This is similar to mojang's SlotCrafting, except it is tailored specifically to be used with the assemblyTable.
+ * NOTE: This will only work correctly if it's used with AssemblyTableRecipes.
+ *
+ * @author PseudonymPatel
+ * @since 2020-8-3
+ */
+public class SlotCraftingResult extends Slot {
+    private final IInventory craftMatrix;
+    private final EntityPlayer player;
+    private int amountCrafted;
+
+    public SlotCraftingResult(EntityPlayer player, IInventory craftingInventory, IInventory inventoryIn, int slotIndex, int xPosition, int yPosition) {
+        super(inventoryIn, slotIndex, xPosition, yPosition);
+        this.player = player;
+        this.craftMatrix = craftingInventory;
+    }
+
+    public boolean isItemValid(ItemStack stack) {
+        return false;
+    }
+
+    public ItemStack decrStackSize(int amount) {
+        if (this.getHasStack()) {
+            this.amountCrafted += Math.min(amount, this.getStack().getCount());
+        }
+
+        return super.decrStackSize(amount);
+    }
+
+    protected void onCrafting(ItemStack stack, int amount) {
+        this.amountCrafted += amount;
+        this.onCrafting(stack);
+    }
+
+    protected void onSwapCraft(int p_190900_1_) {
+        this.amountCrafted += p_190900_1_;
+    }
+
+    protected void onCrafting(ItemStack stack) {
+        if (this.amountCrafted > 0) {
+            stack.onCrafting(this.player.world, this.player, this.amountCrafted);
+            net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerCraftingEvent(this.player, stack, craftMatrix);
+        }
+
+        this.amountCrafted = 0;
+
+        InventoryCraftResult inventorycraftresult = (InventoryCraftResult)this.inventory;
+        IRecipe irecipe = inventorycraftresult.getRecipeUsed();
+
+        //the following can be uncommented later (it all works) if we want to do things with the recipe that was just crafted.
+        //I commented it out because it was adding a recipe to the recipe book for whatever was just crafted, but that should not be done (line 76)
+        if (irecipe != null && !irecipe.isDynamic()) {
+//            this.player.unlockRecipes(Lists.newArrayList(irecipe));
+            inventorycraftresult.setRecipeUsed((IRecipe) null);
+        }
+    }
+
+    public ItemStack onTake(EntityPlayer thePlayer, ItemStack stack) {
```

Comment:
wouldn't it be better to just use a custom Slot class for every block that needs it? This instanceof check could make it look complicated

Reply:
Oh! sorry I forgot about that. I was at one point going to use it for both the train wb and assembly tables, but then started to convert it to be used with the assembly table, and forgot to untrack it from git. It's not used anywhere at the moment, and it'll be a lot different when finished